src/builtins.ts

```diff
@@ -2056,6 +2072,297 @@ function builtin_store(ctx: BuiltinContext): ExpressionRef {
+      return ret;
+    }
+  }
+  compiler.error(
+    DiagnosticCode.Operation_0_cannot_be_applied_to_type_1,
+    ctx.reportNode.typeArgumentsRange,
+    "add",
+    type.toString()
+  );
+  return module.unreachable();
+}
+builtins.set(BuiltinNames.add, builtin_add);
+
+// sub<T?>(left: T, right: T) -> T
+function builtin_sub(ctx: BuiltinContext): ExpressionRef {
+  var compiler = ctx.compiler;
+  var module = compiler.module;
+  if (checkTypeOptional(ctx, true) | checkArgsRequired(ctx, 2))
+    return module.unreachable();
+  var operands = ctx.operands;
+  var typeArguments = ctx.typeArguments;
+  var left = operands[0];
+  var arg0 = typeArguments
+    ? compiler.compileExpression(
+        left,
+        typeArguments[0],
+        Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+      )
+    : compiler.compileExpression(operands[0], Type.auto, Constraints.MUST_WRAP);
+  var type = compiler.currentType;
+  if (type.isValue) {
+    let arg1: ExpressionRef;
+    if (!typeArguments && left.isNumericLiteral) {
+      // prefer right type
+      arg1 = compiler.compileExpression(
+        operands[1],
+        type,
+        Constraints.MUST_WRAP
+      );
+      if (compiler.currentType != type) {
+        arg0 = compiler.compileExpression(
+          left,
+          (type = compiler.currentType),
+          Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+        );
+      }
+    } else {
+      arg1 = compiler.compileExpression(
+        operands[1],
+        type,
+        Constraints.CONV_IMPLICIT | Constraints.MUST_WRAP
+      );
+    }
+    let op: BinaryOp = -1;
+    switch (type.kind) {
+      case TypeKind.I8:
+      case TypeKind.I16:
+      case TypeKind.I32:
+      case TypeKind.U8:
+      case TypeKind.U16:
+      case TypeKind.U32:
+      case TypeKind.BOOL: {
+        op = BinaryOp.SubI32;
+        break;
+      }
+      case TypeKind.I64:
+      case TypeKind.U64: {
+        op = BinaryOp.SubI64;
+        break;
+      }
+      case TypeKind.ISIZE: {
+        op = compiler.options.isWasm64 ? BinaryOp.SubI64 : BinaryOp.SubI32;
+        break;
+      }
+      case TypeKind.USIZE: {
+        op = compiler.options.isWasm64 ? BinaryOp.SubI64 : BinaryOp.SubI32;
+        break;
+      }
+      case TypeKind.F32:
+        return module.binary(BinaryOp.SubF32, arg0, arg1);
+      case TypeKind.F64:
+        return module.binary(BinaryOp.SubF64, arg0, arg1);
+    }
+    if (op != -1) {
+      let flow = compiler.currentFlow;
+      let nativeType = type.toNativeType();
+      let temp1 = flow.getTempLocal(type);
+      flow.setLocalFlag(temp1.index, LocalFlags.WRAPPED);
+      let temp2 = flow.getTempLocal(type);
+      flow.setLocalFlag(temp2.index, LocalFlags.WRAPPED);
+      let ret =  module.binary(
+        op,
+        module.local_get(temp1.index, nativeType),
+        module.local_get(temp2.index, nativeType)
+      );
+      flow.freeTempLocal(temp2);
+      flow.freeTempLocal(temp1);
+      return ret;
```

Comment:
This doesn't seem correct, in that two temporary locals are created, but none is assigned, so chances are that the binary operation doesn't operate on the expected values. Looks like this part of the code is based on the integer polyfills employed in `abs` etc., which is not necessary for add/sub/mul but can be simpler, without temporaries.

Reply:
Yes, it could be similar to approach of handling f32 / f64. Only one difference for `TypeKind.BOOL, TypeKind.I8, TypeKind.U8, TypeKind.U16, TypeKind.I16` group you should wrap result with `ensureSmallIntegerWrap`:

```ts
   switch (type.kind) {
      case TypeKind.I8:
      case TypeKind.I16:
      case TypeKind.U8:
      case TypeKind.U16:
      case TypeKind.BOOL: {
        return compiler.ensureSmallIntegerWrap(
          module.binary(BinaryOp.AddI32, arg0, arg1),
          type
        );
      }
      case TypeKind.I32:
      case TypeKind.U32: 
         return module.binary(BinaryOp.AddI32, arg0, arg1);
      case TypeKind.USIZE:
      case TypeKind.ISIZE: {
        return module.binary(
          compiler.options.isWasm64
            ? BinaryOp.AddI32
            : BinaryOp.AddI64,
          arg0, arg1
        );
      }
      case TypeKind.I64:
      case TypeKind.U64: return module.binary(BinaryOp.AddI64, arg0, arg1);
    }
    // error handling
```

Reply:
Yes, that was that, thank you!

I indeed took it from `builtin_max`

Reply:
See my example above. You should specially handle small integers like `i8` or `u16`.

Reply:
Thank you, removed temporary from add/mul/sub and added `ensureSmallIntegerWrap` to them